Start using Architecture Decision Records (ADRs) in Sourcegraph. (#34511)
sourcegraph/sourcegraph: doc/adr/0001-record-architecture-decisions.md

# 1. Record architecture decisions

Date: 2022-04-26

## Context

We have identified a need to capture architecturally significant decisions that will help us understand how our codebase evolved over time.

- Usage of Architecture Decision Records (ADRs) has been proposed at the [Backend Crew Meeting](https://docs.google.com/document/d/1Y51d863Nuqr9BzbTbwwSF9jes0ro71vB724oc2p8qsQ/edit#heading=h.cwkac4n1yeqr).
- The conversations have been initiated from the Slack thread (where we discussed whether we want to unify our RPC mechanism on gRPC on Protocol Buffers). As a side effect of that conversation, we realized there wasn't a good tool to capture architecturally significant decisions.

## Decision

- We will use Architecture Decision Records (ADRs) to log architecturally significant decisions.

> - ADRs are not meant to replace our current RFC process but to complement it by capturing decisions made in RFCs.

Reviewer: Do you think there's a danger of people trying to skip to the end here and proposing ADRs directly instead of first RFC and then logging decisions in RFC? If so, how would we avoid that?

Author: I think this shouldn't be happening, as ADRs should not serve for proposing things but for recording decisions made elsewhere, be at an RFC, a GitHub issue discussion, or any other relevant place.

We can call such behavior out if we see it.

> - We will use Architecture Decision Records (ADRs) to log architecturally significant decisions.
> - ADRs are not meant to replace our current RFC process but to complement it by capturing decisions made in RFCs.
> - As a rule of thumb, we recommend creating ADRs only for those decisions that have a notable architectural impact on our codebase.

Reviewer 2: ```suggestion
- We will use Architecture Decision Records (ADRs) only for logging decisions that have notable architectural impact on our codebase.
- ADRs are not meant to replace our current RFC process but to complement it by capturing decisions made in RFCs.
```

Here is my attempt regarding my feedback.
